tests/test_data_cloud.py

```diff
@@ -597,3 +598,90 @@ def test(self):
         self.color_patch.start()
         self._test()
         self.color_patch.stop()
+
+
+class TestRecursiveSyncOperations(TestDataCloudBase):
+    def main(self, args):
+        ret = main(args)
+        self.assertEqual(ret, 0)
+
+    def _get_url(self):
+        self.dname = get_local_url()
+        return self.dname
+
+    def _should_test(self):
+        return True
+
+    def _get_cloud_class(self):
+        return RemoteLOCAL
+
+    def _prepare_repo(self):
+        self.main(['remote',
+                   'add',
+                   '-d',
+                   TEST_REMOTE,
+                   self.cloud._cloud.cache_dir])
+        self.create(self.DATA, self.DATA_CONTENTS)
+        self.create(self.DATA_SUB, self.DATA_SUB_CONTENTS)
+
+        self.dvc.add(self.DATA)
+        self.dvc.add(self.DATA_SUB_CONTENTS)
+
+        os.chdir(self._root_dir)
+
+    def _remove_local_data_files(self):
+        os.remove(self.DATA)
+        os.remove(self.DATA_SUB)
+
+    def _test_recursive_pull(self):
+        self._remove_local_data_files()
```

Comment:
Use self._clear_local_cache() here in addition to self._remove_local_data_files(), so that you could make sure that `pull` does both fetches and checks-out everything it needs.